cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java

```diff
@@ -0,0 +1,110 @@
+import co.cask.cdap.templates.etl.api.StageConfigurer;
+import co.cask.cdap.templates.etl.api.batch.BatchSink;
+import co.cask.cdap.templates.etl.api.batch.BatchSinkContext;
+import co.cask.cdap.templates.etl.api.config.ETLStage;
+import co.cask.cdap.templates.etl.common.ETLUtils;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.collect.Maps;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
+import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;
+
+import java.lang.reflect.Type;
+import java.util.Map;
+
+/**
+ * CDAP {@link PartitionedFileSet} Batch Sink
+ */
+public class PartitionedFileSetSink extends BatchSink<Text, Text> {
+
+  private static final String FILESET_NAME = "filesetName";
+  private static final String PARTITION_KEY = "partitionKey";
+  private static final String PARTITIONING = "partitioning";
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(PartitionedFileSetSink.class.getSimpleName());
+    configurer.setDescription("CDAP Partitioned fileSet batch sink");
+    configurer.addProperty(new Property(FILESET_NAME, "Partitioned fileset name", true));
+    configurer.addProperty(new Property(PARTITION_KEY, "Partition key in JSON format to use while writing " +
+      "ex: {\"league\" : \"nfl\", \"season\" : \"1980\"}", true));
+    configurer.addProperty(new Property(PARTITIONING, "Partitioning to use for the Partitioned FileSet in JSON " +
+      "format. Should be provided if you want the pipeline to create the fileset. ex: " +
+      "{\"league\" : \"string\", \"season\" : \"int\"}", false));
+    configurer.addProperty(new Property(TextOutputFormat.SEPERATOR, "The output format separator. Must be provided " +
+      "if you want the pipeline to create the fileset", false));
+  }
+
+  @Override
+  public void configurePipeline(ETLStage stageConfig, PipelineConfigurer pipelineConfigurer) {
+    // if the partitioning is provided then we should try to create the fileset here
+    if (!Strings.isNullOrEmpty(stageConfig.getProperties().get(PARTITIONING))) {
+      String pfsName = stageConfig.getProperties().get(FILESET_NAME);
+      String separator = stageConfig.getProperties().get(TextOutputFormat.SEPERATOR);
+      Preconditions.checkArgument(!Strings.isNullOrEmpty(pfsName), "PartitionedFileSet name must be given.");
+      Preconditions.checkArgument(!Strings.isNullOrEmpty(pfsName), "A separator for " +
+        TextOutputFormat.SEPERATOR + "must be given");
+      pipelineConfigurer.createDataset(pfsName, PartitionedFileSet.class.getName(),
+                                       PartitionedFileSetProperties.builder()
+        .setPartitioning(ETLUtils.createPartitioning(stageConfig.getProperties().get(PARTITIONING)))
+        .setInputFormat(TextInputFormat.class)
```

Comment:
as discussed, these should all be properties that the user sets when creating a pipeline